### DISCUSSION TITLE
Manual tagging of math equations

### DISCUSSION THREAD

**Discussion by juliusross1** [latex3/tagging-project#717] [General]
Is it possible with this tagging project to allow manual creation of the alt tags for mathematics equations?  Something like $$a^2 + b^2 = c^2 \Description{a squared plus b squared equals c squared}$$

Thanks!
JR

**Comment by FrankMittelbach** [comment]
As of today, the answer is *no*.

However, in a later phase of the project all environments (and commands) that could resonably have an `alt` text set will support a first optional argument in which this alternative text can be specified. E.g.,
```
\begin{equation}[alt=a squared plus b squared equals c squared]
  a^2 + b^2 = c^2
\end{equation}
```
Environments or commands that already have a first optional argument, will get an extended/enhanced version of that optional argument in which it will also be possible to specify alt text as well as whatever the original optinal argument was used for.

However, with the automatic generation of MathML for the formula content, which is already now available, and with suitable readers that understand and process MathML, (e.g., the newest version of FoxIt) you will find that the automation is probably getting you better results without the need to write manual alt texts.

For inline math, i.e., `$...$` we haven't decided how alt text could be specified, maybe only via the environment form `\begin{math}...` as in most case it should not be needed.

**Reply by juliusross1** [reply]
First, thank you for the quick answer as well as your fantastic work on this it looks really good.

You are likely already aware, but updates to ADA rules in the US mean that universities here are looking at a hard April 2026 deadline to make all student facing documents meet some specific accessibility standards.  So I expect there will be a spike in interest in your project very soon if not already.   At least from what I can see, having the ability to manually add alt to displayed equations would be enough for our needs (in addition to what you already have done).    Further there may be issues with requiring our students to use latest versions of pdf readers; I am not sure what the expectation is here.  But I 100% appreciate your project is multiyear and I do not expect it to fit anybody else's timetable!

On another point, I am curious how the MathML solution might end up looking in the end.  For if I write $\mu(X)$ in the source and ask a pdf reader to read it out loud it could be read as "mu open bracket X close bracket" or "mu of X"  even perhaps "the mean of X" if context dictates that is what "mu" means.

**Comment by u-fischer** [comment]
There is already support for alt text on math. You can activate an automatic alt text with `\tagpdfsetup{math/alt/use}` (the alt text is then basically the LaTeX source). (

We are currently disabling the alt text as it previously "hid" the attached mathml in foxit but this seems to have changed in yesterdays update of foxit. If this is confirmed we could change the setting again and activate the alt by default again. 

It is also not very difficult to add some support for a manual text for display math. The main problem is to enhance all environments to take an optional argument, but as a first step one could simply define a command that must be issued before. 

~~~~
\DocumentMetadata{pdfversion=2.0,pdfstandard=ua-2,testphase={phase-III,math}} 
\documentclass{book}
\tagpdfsetup{math/alt/use} %activate alt test.
\ExplSyntaxOn
\NewDocumentCommand\addmathalttext{mm}
{\AddToHookNext{env/#1/begin}{\tl_set:Nn \l__math_content_template_tl
   {#2}}}
\ExplSyntaxOff
\begin{document}
$1=1$ 

\addmathalttext{equation}{Some description of the equation}
\begin{equation}
a=b
\end{equation}

\begin{equation}
x=y
\end{equation}

\end{document}
~~~~
 

Please note that the requirement for an alt text stems from the times before mathml when formulas were inserted as images and so needed some description. One can see the difference if one compare the UA-1 and UA-2 standards:

UA-1 writes basically only two sentences about math. It starts with
> All mathematical expressions shall be enclosed within a Formula tag as detailed in ISO 32000-1:2008, 14.8.4.5 and shall have an Alt attribute

and UA-2 writes a much longer section which starts with 
> Mathematical expressions shall be represented using either structure types defined in presentation MathML, as specified in ISO 32000-2:2020, 14.8.6.3, or as presentation MathML via the Associated Files mechanism specified in ISO 32000-2:2020, 14.13.6, or both.


So if you adhere to UA-2 mathml is the way to go. 

> On another point, I am curious how the MathML solution might end up looking in the end.

The idea is that it should be possible to add an intent to the mathml and that the AT (MathCat) picks that up.

**Comment by davidcarlisle** [comment]
>  At least from what I can see, having the ability to manually add alt to displayed equations would be enough for our needs 

adding a text fallback is of course better than nothing but gives a poorer experience than providing the mathml structure (assuming suitable software) notably given a full text string the reader can not navigate the structure, re-reading subterms etc, also it is not possible to generate correct Braille, mathematical braille is not just the braille of an english text reading.

> I am curious how the MathML solution might end up looking in the end. For if I write `\mu(x)` in the source and ask a pdf reader to read it out loud it could be read as "mu open bracket X close bracket" or "mu of X" " even perhaps "the mean of X" if context dictates that is what "mu" means.


all three of those are possible:

If the associated mathml is  `<mi>μ</mi><mo>(</mo><mi>x</mi><mo>)</mo>` it will be `mu of x` by default , with something closer to `mu open bracket x close bracket` if reading in character-by-character mode (the reader's option). If the associated mathml has `intent` eg `<mi intent="mean">μ</mi><mo>(</mo><mi>x</mi><mo>)</mo>` the default reading (in english)  would be `mean of x` 

Currently there isn't an automatic way of generating `intent` annotations from tex but that will come, and the system is designed to allow hand annotation where needed. In the end you should be able to define document commands such as `\mean{x}` that produces tagging that makes fully annotated mathml.

The first example at https://github.com/latex3/tagging-project/discussions/72  has the mathml annotated 

![image](https://github.com/user-attachments/assets/e66c61ab-b4a3-464b-8e87-fe23ee58be0d)


is associated to

```
<math>
 <msqrt><msup><mi>x</mi><mn>2</mn></msup></msqrt>
 <mo>=</mo>
 <mrow intent="absolute-value($x)"><mo>|</mo><mi arg="x">x</mi><mo>|</mo></mrow>
</math>
```

and reads as

```
the square root of x squared equals the absolute value of x
```

using the foxit pdf reader with NVDA/MathCat screen reader.
